cmd/sovereignnode/dataCodec/dataCodec.go

```diff
@@ -0,0 +1,482 @@
+	return &DataCodec{
+		Serializer: abi.NewSerializer(codec),
+		Marshaller: marshaller,
+	}, nil
+}
+
+// SerializeEventData will receive an event data and serialize it
+func (dc *DataCodec) SerializeEventData(eventData sovereign.EventData) ([]byte, error) {
+	var gasLimit any
+	var function any
+	var args any
+	if eventData.TransferData != nil {
+		gasLimit = abi.U64Value{Value: eventData.GasLimit}
+		function = abi.BytesValue{Value: eventData.Function}
+		arguments := make([]any, len(eventData.Args))
+		for i, arg := range eventData.Args {
+			arguments[i] = abi.BytesValue{Value: arg}
+		}
+		args = abi.InputListValue{
+			Items: arguments,
+		}
+	} else {
+		gasLimit = nil
+		function = nil
+		args = nil
+	}
+
+	eventDataStruct := abi.StructValue{
+		Fields: []abi.Field{
+			{
+				Name:  "tx_nonce",
+				Value: abi.U64Value{Value: eventData.Nonce},
+			},
+			{
+				Name: "gas_limit",
+				Value: abi.OptionValue{
+					Value: gasLimit,
+				},
+			},
+			{
+				Name: "function",
+				Value: abi.OptionValue{
+					Value: function,
+				},
+			},
+			{
+				Name: "args",
+				Value: abi.OptionValue{
+					Value: args,
+				},
+			},
+		},
+	}
+
+	encodedOp, err := dc.Serializer.Serialize([]any{eventDataStruct})
+	if err != nil {
+		return nil, err
+	}
+
+	operationBytes, _ := hex.DecodeString(encodedOp)
+	return operationBytes, nil
+}
+
+// DeserializeEventData will deserialize bytes to an event data structure
+func (dc *DataCodec) DeserializeEventData(data []byte) (*sovereign.EventData, error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("empty data provided")
+	}
+
+	eventDataStruct := abi.StructValue{
+		Fields: []abi.Field{
+			{
+				Name:  "tx_nonce",
+				Value: &abi.U64Value{},
+			},
+			{
+				Name: "gas_limit",
+				Value: &abi.OptionValue{
+					Value: &abi.U64Value{},
+				},
+			},
+			{
+				Name: "function",
+				Value: &abi.OptionValue{
+					Value: &abi.BytesValue{},
+				},
+			},
+			{
+				Name: "args",
+				Value: &abi.OptionValue{
+					Value: &abi.OutputListValue{
+						ItemCreator: func() any { return &abi.BytesValue{} },
+					},
+				},
+			},
+		},
+	}
+
+	err := dc.Serializer.Deserialize(hex.EncodeToString(data), []any{&eventDataStruct})
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := eventDataStruct.Fields[0].Value.(*abi.U64Value).Value
+
+	gasLimit := uint64(0)
+	optionGas := eventDataStruct.Fields[1].Value.(*abi.OptionValue).Value
+	if optionGas != nil {
+		gasLimit = optionGas.(*abi.U64Value).Value
+	}
+
+	function := make([]byte, 0)
+	optionFunc := eventDataStruct.Fields[2].Value.(*abi.OptionValue).Value
+	if optionFunc != nil {
+		function = optionFunc.(*abi.BytesValue).Value
+	}
+
+	args := make([][]byte, 0)
+	optionArgs := eventDataStruct.Fields[3].Value.(*abi.OptionValue).Value
+	if optionArgs != nil {
+		items := optionArgs.(*abi.OutputListValue).Items
+		if items != nil && len(items) > 0 {
+			for _, item := range items {
+				arg := item.(*abi.BytesValue).Value
```

Comment:
@andreibancioiu can you answer this question please? :smile:

Reply:
As written, no. 

However, you should not trust that the codec is bug-free: e.g. check length of `eventDataStruct.Fields` before attempting to access elements by index; return error instead of panicking if a cast fails (even though it should not).

Reply:
change the implementation for `deserialize` functions based on Andrei's comment